File: api/src/models/baseline.py
"""Pydantic models for baseline snapshots (Piece 50)."""
from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, Field, conint


# Use a plain conint to keep stat values non-negative — the bridge
# could in principle send -1; we'd rather 422 than store nonsense.
NonNeg = Annotated[int, Field(ge=0)]


class BaselineFrameworkStats(BaseModel):
    """Per-framework slice of the aggregate. The keys are framework
    buckets (e.g. 'iso27001', 'pci_dss') — same name space as
    customer_policies.framework_bucket."""
    passing: NonNeg
    failing: NonNeg


class BaselineSummary(BaseModel):
    """The aggregate evaluation stats the bridge POSTs. Schema is the
    bridge's contract; the portal stores it opaque after Pydantic
    validation."""
    host_count: NonNeg
    total_evaluations: NonNeg
    passing: NonNeg
    failing: NonNeg
    errors: NonNeg = 0
    by_framework: dict[str, BaselineFrameworkStats] = Field(
        default_factory=dict,
        description=(
            "Per-framework breakdown. Optional — bridge can omit if it "
            "only has a global view."
        ),
    )


class BaselineIngestRequest(BaseModel):
    """Body the bridge POSTs to /tenants/{id}/baselines.

    `source` is set server-side to 'bridge_push' for this endpoint;
    a separate operator-side route can post with `source='manual'`.
    """
    bundle_sha256: str = Field(..., min_length=64, max_length=64)
    summary: BaselineSummary
    label: str | None = Field(default=None, max_length=200)


class BaselineSnapshotSummary(BaseModel):
    """List-view row. Omits the full `summary` jsonb to keep the
    payload light for the history table. The detail endpoint
    returns the full thing."""
    id: UUID
    tenant_id: UUID
    bundle_sha256: str
    captured_at: datetime
    captured_by_email: str | None = None
    label: str | None = None
    source: Literal["bridge_push", "manual", "scheduled"]
    host_count: int
    passing: int
    failing: int


class BaselineSnapshotDetail(BaselineSnapshotSummary):
    """Detail view — includes the full BaselineSummary, with its
    by_framework breakdown."""
    summary: BaselineSummary

File: frontend/src/types/baseline.ts
// Mirrors api/src/models/baseline.py.
//
// A baseline snapshot is one tenant's compliance evaluation state at a
// point in time. The bridge POSTs them automatically; an operator can
// also manually import one via the UI for backfills or testing.

export type BaselineSource = "bridge_push" | "manual" | "scheduled";

export interface BaselineFrameworkStats {
  passing: number;
  failing: number;
}

export interface BaselineSummary {
  host_count: number;
  total_evaluations: number;
  passing: number;
  failing: number;
  errors: number;
  by_framework: Record<string, BaselineFrameworkStats>;
}

export interface BaselineIngestRequest {
  bundle_sha256: string;
  summary: BaselineSummary;
  label?: string | null;
}

// List row — lean. summary jsonb is omitted; we project the most-used
// counters out of it server-side so the list payload stays small.
export interface BaselineSnapshotSummary {
  id: string;
  tenant_id: string;
  bundle_sha256: string;
  captured_at: string;
  captured_by_email: string | null;
  label: string | null;
  source: BaselineSource;
  host_count: number;
  passing: number;
  failing: number;
}

export interface BaselineSnapshotDetail extends BaselineSnapshotSummary {
  summary: BaselineSummary;
}
